### ISSUE TITLE
Activity Marker

### ISSUE THREAD

**Issue by awkay** [fulcrologic/fulcro#242]
Currently only loads manipulate an activity marker.  Mutations are required to do their own manipulation.  This made sense initially because mutations are optimistic by default, and additional work must be done to deal with mutation return values, etc.  So, the intention was that you'd implement your own mutation activity process.

Wilker's proposal in incubator has some nice aspects in terms of giving an official lifecycle to a mutation, but it doesn't cover the "easy case" of just seeing that things are "in motion" on the network.

I was able to add an activity tracking into Fulcro reconciler's internal state with commit c291efebc2da12145f8d21b229f2ec4d23279338
which simply adds an internal map to the reconciler that can be accessed via `fulcro.client.impl.protocols/active-remotes`.

I'm considering what the reasonable approach is:

- Add a new key to top-level state that indicates "any network activity"
- Make the global loading activity marker really *mean* network activity, whcih is what many ppl expect.
- Add a `primitives/network-active?` function that returns true/false

The other motivating case at the moment is claudiu's attempt to do SSR with node.  He wants a way to get some kind of async notification that the network has become idle.  The fact that reconciler's state is an atom (not app state, reconciler state), means that a watch can be attached with the current scheme like so:

```
                :started-callback (fn [{:keys [reconciler] :as started-app}]
                                    (let [trigger (debounce (fn notify-idle-network [active-remotes]
                                                              (if (empty? active-remotes)
                                                                (js/console.log :NETWORK-IDLE))) 2)]
                                      (add-watch (:state reconciler) :active-remotes
                                        (fn [_ _ old-state new-state]
                                          (when (not= (:active-remotes old-state) (:active-remotes new-state))
                                            (trigger (:active-remotes new-state)))))))
```

The debounce is necessary because there will be rapid done/start edges if there are multiple requests queued. 2ms should be plenty, since really it's just giving up the CPU long enough for the next request to start.

This could be turned into an API call that would encompass this general facility, like this:

```
(defn notify-network-idle! [reconciler callback timeout] ...)
```

and the notification could even remove the watch for a single-shot notification.  The only use-case I see for this is node.js SSR, but it is a clearly useful thing for that.

I'd be interested in feedback.

**Comment by awkay** [comment]
I've added a `prim/notify-network-idle!` in this commit:

ac1b3d66a2dc10cada995d3e404512786839d2f1

It codifies a single-shot way to ask for this.

**Comment by claudiuapetrei** [comment]
I also have another use-case for this. It's related to the routing approach that I'm building for a project, and was hoping that I can make it friendly enough that it could make it's way into fulcro or as a standalone library,

I've got a basic working example at https://5-dot-infodogs-com.appspot.com/

The main ideas are:
- Co-locate route-data load on component (pobably macro over defsc that adds a protocol)
- When you change the page: 
  * stay on current route & show loading marker 
  * router calls component.onload and changes the screen after the route data is populated
- dynamic router (set-query!) that support code-splitting at module level (not route, ex: private & public bundles)
- handles clicking on a slow loading route and the switching to a route that loads instantly since it has no on-load
- handle the html5 routing out of the box
- should require minimal setup to have everything working

For it to be a nice implementation I would really like to treat the on-load function as a black box and the router should just look at the network loading maker to see when it has finished so it can change the screen. 

Routing with fulcro was one of my major pain-points when getting started, still feel the routing story is not as great as it could be. I really like pre-loading data and then changing the route since I think it's a nicer UI experience and reduces a lot of work since you don't need all those loading states for components on page initial load.

Think it would be nice to have the option to just do something like and just have everything working:
```
(fc/new-fulcro-client
               :started-callback (fn [{:keys [reconciler]}]
                                              (init-router reconciler options)     ))```
```
(defscpage BlogPosts [this {:keys [person/name person/age]}]
  {:query         ...
   :initial-state ...
   :on-load  (fn [bidi-route-info] 
                    (df/load ...)
                    (when x 
                      (df/load ...)) 
                    (prim/transact! ...))
   (dom/div ...)))```

This also builds on the SSR story for node.js. Since then SSR could be as simple as call component.load, then call a function that checks network activity, builds and returns the state.

**Comment by awkay** [comment]
This really needs more thought.  The new function is workable, but not ideal.  I think perhaps an additional option to `new-fulcro-client` like `:after-initial-load` is a better fit for the specific SSR case, and a separate issue is the need for a general networking activity marker.  

Not sure when I'll get to those.  If you're working with a single remote, then a simple workaround is to have your SSR response wait on a channel, and your final mutation put a value into that channel.  You could coordinate multiple remotes this way as well. 

I don't really want to add what I've done to official API right now, as it doesn't solve any case widely enough and well enough.

**Comment by awkay** [comment]
I think I'm going to close this one.  It is relatively easy, I think, to get the SSR for node working with core.async, and I'm not ready to add additional activity marker support at this time.  Open to contribution.
